core/src/term/mod.rs

```diff
@@ -261,6 +261,13 @@ pub enum Term {
     ///
     /// This is a temporary solution, and will be removed in the future.
     Closure(CacheIndex),
+
+    #[serde(skip)]
+    /// An opaque value that cannot be constructed within Nickel code.
```

Comment:
Could be worth to insist on the fact that they cannot be constructed, but also should _never_ be observable from within Nickel code (compared or distinguished by any means).

Reply:
I made sure they can't be compared for equality; is there some other kind of observability that needs to be handled?

I was imagining that these values could be copied around (and possibly manipulated by functions provided by the nickel embedder). So you'd be able to write a contract like `{ username | String, token | ForeignId }`, and then by applying this contract you could observe the presence or absence of the token.

Reply:
Right, I said observable, but the right technical term is probably _separable_, which means "distinguishable".

That is, for every context `C` (that you can think of just as a function here), for any pair of foreign keys `k1` and `k2`, then if `C k1 k2 ~> v` (the context evaluates to the value `v`), then for all other foreign keys `k` and `l`, `C k l ~> v'` with `v ~ v'` (I won't define the precise meaning of `~`, but let's say it's an observational equivalence). Maybe we need to extend that to arbitrarily long finite lists of key.

Put differently, no result of an expression can depend on the actual values of the keys. They must be all interchangeable without affecting the semantics. We need to include error messages as well (so my above specification is incomplete), because that's another way for a malicious user to get the value of the key indirectly - but you properly don't print the content of the key in the pretty printer, which is all good.

Reply:
comparing by equality is actually specifically fine and useful. Even being able to use them as keys in maps is fine and useful as long as that doesn't reveal the content of the opaque reference directly or indirectly. (though there is some subtlety of exactly which equality is being checked, since there could be two copies of the same reference, two different references to the same local data, or two different references to promises which will eventually become the same)
It's being able to forge them or inspect the backing data that's the problem, and as long as there's no way to make a new opaque reference (from inside the language; making them from foreign functions into the trusted platform is fine) that is satisfied, and no way to look up what's behind the foreign reference. Being able to check the arbitrary number in the foreign reference that indexes the backing table isn't even a security issue as long as the table itself is impossible to access, but it is good practice to forbid that since any kind of behavioral reliance on the specific values of those numbers is fragile and should be prevented, and if the numbers are viewable in the first place someone accidentally allowing forgery is more likely.

Reply:
I think using them as map keys is much more invasive. Currently record fields in Nickel are all strings, that you can list with stdlib functions. So opaque value would need special casing here.

Regarding equality, this is easy to add - I just wonder if there are other use-cases where you would want to not even have equality. In any case, I would propose to start the most restrictive possible and move forward with this PR, and then see cases by case what we would add and why, if that sounds good to you all.